### PR TITLE
Add PageConvention overloads for areas

### LINKED ISSUE #7246: Add overloads to PageConventions to support areas
`[Page|Folder]RouteModelConvention` and `[Page|Folder]ApplicationModelConvention` work by matching `ViewEnginePath` with no support for additional filtering. To make these conventions work for areas, we need to accept an additional parameter the "area" that filters models that belong the specified area. 

Here's the additional set of public API we would need:

* AddAreaPageRoute
* AllowAnonymousToAreaFolder
* AllowAnonymousToAreaPage
* AuthorizeAreaFolder
* AuthorizeAreaPage

We could also add overloads to our existing methods to support passing in this additional value. e.g. ` AddPageRoute(string areaName, string pageName, string route)` although the order in which parameters appear makes this less appealing.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #7246

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs [hunk 1]
```diff
@@ -67,6 +95,34 @@ public IPageApplicationModelConvention AddFolderApplicationModelConvention(strin
             return Add(new FolderApplicationModelConvention(folderPath, action));
         }
 
+        /// <summary>
+        /// Creates and adds an <see cref="IPageApplicationModelConvention"/> that invokes an action on
+        /// <see cref="PageApplicationModel"/> instances for all pages under the specified area folder.
+        /// </summary>
+        /// <param name="areaName">The name of area.</param>
+        /// <param name="folderPath">The path of the folder relative to the Razor Pages root. e.g. <c>/Users/</c></param>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs]: This is relative to the area root, not the pages root.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs [hunk 2]
```diff
@@ -105,6 +186,31 @@ public IPageRouteModelConvention AddFolderRouteModelConvention(string folderPath
             return Add(new FolderRouteModelConvention(folderPath, action));
         }
 
+        /// <summary>
+        /// Creates and adds an <see cref="IPageRouteModelConvention"/> that invokes an action on
+        /// <see cref="PageRouteModel"/> instances for all page under the specified area folder.
+        /// </summary>
+        /// <param name="areaName">The area name.</param>
+        /// <param name="folderPath">The path of the folder relative to the Razor Pages root. e.g. <c>/Users/</c></param>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs]: relative to the area root

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModel.cs [hunk 3]
```diff
@@ -21,9 +21,21 @@ public class PageRouteModel
...
         public PageRouteModel(string relativePath, string viewEnginePath)
+            : this(relativePath, viewEnginePath, areaName: null)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="PageRouteModel"/>.
+        /// </summary>
+        /// <param name="relativePath">The application relative path of the page.</param>
+        /// <param name="viewEnginePath">The path relative to the base path for page discovery.</param>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModel.cs]: This is also relative to the area root right?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/PageConventionCollectionExtensions.cs [hunk 4]
```diff
@@ -105,6 +138,39 @@ public static PageConventionCollection AllowAnonymousToFolder(this PageConventio
             return conventions;
         }
 
+        /// <summary>
+        /// Adds a <see cref="AllowAnonymousFilter"/> to all pages under the specified area folder.
+        /// </summary>
+        /// <param name="conventions">The <see cref="PageConventionCollection"/> to configure.</param>
+        /// <param name="areaName">The area name.</param>
+        /// <param name="folderPath">The folder path relative to <see cref="RazorPagesOptions.RootDirectory"/> inside the
+        /// area directory e.g. <c>/Users/Index</c>.</param>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/PageConventionCollectionExtensions.cs]: Yay

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/PageConventionCollectionExtensions.cs [hunk 5]
```diff
@@ -138,6 +204,41 @@ public static PageConventionCollection AuthorizePage(this PageConventionCollecti
         public static PageConventionCollection AuthorizePage(this PageConventionCollection conventions, string pageName) =>
             AuthorizePage(conventions, pageName, policy: string.Empty);
 
+        /// <summary>
+        /// Adds a <see cref="AuthorizeFilter"/> with the specified policy to the page with the specified name.
+        /// </summary>
+        /// <param name="conventions">The <see cref="PageConventionCollection"/> to configure.</param>
+        /// <param name="areaName">The area name.</param>
+        /// <param name="pageName">The page name. This value is the relative path to <see cref="RazorPagesOptions.RootDirectory"/>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/PageConventionCollectionExtensions.cs]: ohno

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/PageConventionCollectionExtensions.cs [hunk 6]
```diff
@@ -138,6 +204,41 @@ public static PageConventionCollection AuthorizePage(this PageConventionCollecti
...
+        /// <param name="areaName">The area name.</param>
+        /// <param name="pageName">The page name. This value is the relative path to <see cref="RazorPagesOptions.RootDirectory"/> 
+        /// inside the area directory e.g. <c>/Manage/Users</c>. <seealso cref="PageRouteModel.ViewEnginePath" />.</param>
+        /// <param name="policy">The authorization policy.</param>
+        /// <returns>The <see cref="PageConventionCollection"/>.</returns>
+        public static PageConventionCollection AuthorizeAreaPage(
+            this PageConventionCollection conventions,
+            string areaName,
+            string pageName,
+            string policy = "")
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/PageConventionCollectionExtensions.cs]: We don't really do default argument values. Just overload it

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/PageConventionCollectionExtensions.cs (hunk 5 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/PageConventionCollectionExtensions.cs]: OH yeah you're right it is page name so it is root relative.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs (hunk 1 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs]: Yeah, these are bad docs. I'll tidy this up

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModel.cs (hunk 3 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModel.cs]: Yup. I'll expand on what this is and say that this is the equivalent of pageName

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/PageConventionCollectionExtensions.cs (hunk 6 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/PageConventionCollectionExtensions.cs]: 😿

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModel.cs (hunk 3 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModel.cs]: Added docs as part of the property description.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageApplicationModel.cs [hunk 7]
```diff
@@ -72,9 +72,22 @@ public PageApplicationModel(PageApplicationModel other)
 
         /// <summary>
         /// Gets the path relative to the base path for page discovery.
+        /// <para>
+        /// This value is the path of the file without extension, relative to the pages root directory.
+        /// e.g. the <see cref="ViewEnginePath"/> for the file /Pages/Catalog/Antiques.cshtml is <c>/Catalog/Antiques</c>
+        /// </para>
+        /// <para>
+        /// In an area, this value is the path of the file without extension, relative to the pages root directory for the specified area.
+        /// e.g. the <see cref="ViewEnginePath"/>  for the file Areas/Identity/Pages/Manage/Accounts.cshtml, is <c>/Manage/Accounts</c>.
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageApplicationModel.cs]: Nice 👍

**Comment by rynowak** [comment]
>  I don't have enough context on how areas work on pages, but maybe there is a way to make it more of a detail like in MVC?

It's a noble thought, but I think areas in pages **must** be fundamental, or not exist at all.

Areas for routing are just data, but that was never the interesting part. Areas are most relevant for view searches, which is central to how pages work. Areas + file paths leads to having some prescriptive ideas about how things need to be laid out.

If you've got ideas about how to improve this I'm interested, but I don't have a big concern to try and push areas to the background for pages. I think it makes much more sense for pages than controllers.

**Pull request closed without merging**